### PR TITLE
init: s/ARCH/EARLY, call it just before arch kernel init

### PR DESCRIPTION
**Pull request by gmarull**
The `ARCH` init level was added to solve a specific problem, call init code (SYS_INIT/devices) before `z_cstart` in the `intel_adsp` platform. The documentation claims it runs before `z_cstart`, but this is only true if the SoC/arch takes care of calling:

```c
z_sys_init_run_level(_SYS_INIT_LEVEL_ARCH);
```

Which is only true for `intel_adsp` nowadays. So in practice, we now have a platform specific init level. This patch proposes to do things in a slightly different way. First, level name is renamed to `EARLY`, to emphasize it runs in the early stage of the boot process. Then, it is handled by the Kernel (inside `z_cstart()` before calling `arch_kernel_init()`). This means that any platform can now use this level. For `intel_adsp`, there should be no changes, other than `gcov_static_init()` will be called before (I assume this will allow to obtain coverage for code called in EARLY?).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

### CODE REVIEW

**Comment by nashif** [comment]
> Which is only true for `intel_adsp` nowadays. So in practice, we now have a platform specific init level.

that is not true, first user indeed, but not the last. This can also be used for X86 for early serial for example and potentially other places where we hard-code initialization instead of using SYS_INIT.

I am fine with name change however. ARCH was chosen because it is invoked in the code path usually implemented by the architecture.

**Review by nashif — APPROVED** [review]

**Comment by gmarull** [comment]
> > Which is only true for `intel_adsp` nowadays. So in practice, we now have a platform specific init level.
> 
> that is not true, first user indeed, but not the last. This can also be used for X86 for early serial for example and potentially other places where we hard-code initialization instead of using SYS_INIT.
> 

The problem is that a device can be generic for multiple architectures (if they share the same IP). So, the init levels should be supported regardless of the architecture, I think. This is what this PR tries to achieve by moving the responsibility of calling the `EARLY` devices to the Kernel.

> I am fine with name change however. ARCH was chosen because it is invoked in the code path usually implemented by the architecture.

**Review by dcpleung — DISMISSED** [review]

**Review by nashif — CHANGES_REQUESTED** [review]
> This is what this PR tries to achieve by moving the responsibility of calling the `EARLY` devices to the Kernel.

ok, did not notice that. I need to check what this entails. This is supposed to be just at the right spot in the architecture boot process and moving it to init.c defeats the original purpose, i.e. console access as early as possible on some architectures/platforms.

**Comment by nashif** [comment]
You missed 2 spots
```
diff --git a/soc/xtensa/intel_adsp/common/boot_complete.c b/soc/xtensa/intel_adsp/common/boot_complete.c
index 98aac9e315..9e5cb043ba 100644
--- a/soc/xtensa/intel_adsp/common/boot_complete.c
+++ b/soc/xtensa/intel_adsp/common/boot_complete.c
@@ -29,4 +29,4 @@ int boot_complete(const struct device *d)
        return 0;
 }

-SYS_INIT(boot_complete, ARCH, CONFIG_KERNEL_INIT_PRIORITY_DEVICE);
+SYS_INIT(boot_complete, EARLY, CONFIG_KERNEL_INIT_PRIORITY_DEVICE);
diff --git a/soc/xtensa/intel_adsp/common/mem_window.c b/soc/xtensa/intel_adsp/common/mem_window.c
index 01cdd6ae70..32d3c32089 100644
--- a/soc/xtensa/intel_adsp/common/mem_window.c
+++ b/soc/xtensa/intel_adsp/common/mem_window.c
@@ -46,7 +46,7 @@ __imr int mem_win_init(const struct device *dev)
                .mem_base = DT_REG_ADDR(DT_PHANDLE(MEM_WINDOW_NODE(n), memory)) + WIN_OFFSET(n),   \
                .initialize = DT_PROP(MEM_WINDOW_NODE(n), initialize),                             \
        };                                                                                         \
-       DEVICE_DT_DEFINE(MEM_WINDOW_NODE(n), mem_win_init, NULL, NULL, &mem_win_config_##n, ARCH,  \
+       DEVICE_DT_DEFINE(MEM_WINDOW_NODE(n), mem_win_init, NULL, NULL, &mem_win_config_##n, EARLY,  \
                         CONFIG_KERNEL_INIT_PRIORITY_DEFAULT, NULL);

 #if DT_NODE_HAS_STATUS(MEM_WINDOW_NODE(0), okay)
```

**Comment by gmarull** [comment]
> > This is what this PR tries to achieve by moving the responsibility of calling the `EARLY` devices to the Kernel.
> 
> ok, did not notice that. I need to check what this entails. This is supposed to be just at the right spot in the architecture boot process and moving it to init.c defeats the original purpose, i.e. console access as early as possible on some architectures/platforms.

“As early as possible” means C domain is reasy? By pushing this at the early part of z_cstart we make sure we are at least in the C domain. And these devices are initialized before anything else like Kernel arch code, so we can offer certain guarantees. In the end this is what intel_adsp does…

**Comment by nashif** [comment]
> “As early as possible” means C domain is ready? By pushing this at the early part of z_cstart we make sure we are at least in the C domain. And these devices are initialized before anything else like Kernel arch code, so we can offer certain guarantees. In the end this is what intel_adsp does…

Yes, this works fine here, I was aiming for more flexibility of where this can be placed in the architecture code. We can revisit if the need arises.

**Review by nashif — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

File: kernel/init.c [hunk 1]
```diff
@@ -427,6 +427,9 @@ FUNC_NORETURN void z_cstart(void)
 	/* gcov hook needed to get the coverage report.*/
 	gcov_static_init();
 
+	/* initialize early init calls */
+	z_sys_init_run_level(_SYS_INIT_LEVEL_EARLY);
+
 	/* perform any architecture-specific initialization */
 	arch_kernel_init();
```
Comment (andyross) [inline kernel/init.c:434]: Given that these two APIs are doing broadly similar things, we might consider deprecating `arch_kernel_init()` and making the existing users use a SYS_INIT() instead.

**Review by peter-mitsis — APPROVED** [review]

File: kernel/init.c (hunk 1 above)
Comment (gmarull) [reply kernel/init.c:434]: I think that makes sense

**Review by stephanosio — APPROVED** [review]

**Pull request merged**